### DISCUSSION TITLE
convert svg to png with size parameter works on windows but causes problems on linux

### DISCUSSION THREAD

**Discussion by comsyspro** [ImageMagick/ImageMagick#3806] [Digital Image Processing]
with the following command i'm converting a svg to a png image with a width size of 2000px
`convert -size 2000x  image.svg image.png`

on windows this command works fine and the converted png image looks good with 2000px width. but on linux (ubuntu 20.04) the same command returns a bad looking png and also doesn't scale the image to the desired width. so where is the problem here? any solution?

**Comment by fmw42** [comment]
I do not believe that -size will affect the SVG rendering. You need to set the -density. SVG is a vector format. You should define the density that will rasterize to the pixel size you want. The file should have the width and height defined inside it for the default density. Changing the density provided in the command line will adjust the rasterized output size.

Please post a link to your SVG file and we can help further.

**Reply by dlemstra** [reply]
The `-size` option can be used when the `SVG` delegate is `RSVG`. That library allows us to specify the size of the image before it is rendered.

**Reply by fmw42** [reply]
Dirk, Is that only for RSVG or does -size work for Inkscape and does it work for MSVG/XML?

**Reply by dlemstra** [reply]
It only works for RSVG.

**Comment by snibgo** [comment]
@comsyspro: please say what versions(s) of IM you use on Windows and Linux.

**Comment by comsyspro** [comment]
now in windows i saw that in delegates there is rsvg listed and in ubuntu this rsvg is missing. 

i also found something here:
https://gist.github.com/maxivak/1476f7e979879da9f75371a86d5627b5
and you should compile imagemagick from source with the option **--with-rsvg=yes**, that is what i will try now.

windows:
Version: ImageMagick 7.0.10-16 Q16 x64 2020-05-31 http://www.imagemagick.org
Copyright: Copyright (C) 1999-2018 ImageMagick Studio LLC
License: http://www.imagemagick.org/script/license.php
Visual C++: 180040629
Features: Cipher DPC Modules OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg lcms lqr lzma openexr pangocairo png ps raw rsvg tiff webp xml zlib

ubuntu installed via apt install imagemagick:
Version: ImageMagick 6.9.10-23 Q16 x86_64 20190101 https://imagemagick.org
Copyright: © 1999-2019 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC Modules OpenMP
Delegates (built-in): bzlib djvu fftw fontconfig freetype jbig jng jpeg lcms lqr ltdl lzma openexr pangocairo png tiff webp wmf x xml zlib

**Reply by fmw42** [reply]
ImageMagick can use any of 3 different SVG renderers:  Inkscape, RSVG delegate, or its own default MSVG/XML renderer. If RSVG is installed it will be used if it shows in the list of delegates. If Inkscape is installed on the system and can be found in your PATH, then it will be used in place of either of the other two, even if they are installed. Inkscape is generally the best, followed by RSVG. Imagemagick's MSVG is the most primitive.  So some SVG file do not do well with MSVG. You would be best to install Inkscape.

**Reply by comsyspro** [reply]
ok, when i install inkscape do i have to set configure variables or somthing else so that imagemagick will get the path from inkscape, or is this found automatically by imagemagick?

here is a link to a simple svg rectangle, which is failing to convert svg to png on linux:
https://we.tl/t-68er1idMXo

**Comment by snibgo** [comment]
As fmw42 says, if Inkscape is on the system path, IM shold find it. Use "-verbose" to verify that IM calls Inkscape.

The "-size" setting, used before reading the SVG, can be used to set the output size. For example:
```
f:\web\im>%IMG7%magick drop.svg info:
drop.svg SVG 500x500 500x500+0+0 16-bit sRGB 339B 0.047u 0:01.062

f:\web\im>%IMG7%magick -size 3000x drop.svg info:
drop.svg SVG 3000x3000 3000x3000+0+0 16-bit sRGB 339B 1.094u 0:02.167
```
This is using IM v7.0.8-64, and Inkscape as the rasterizer. It doesn't work with the MSVG rasterizer. I don't know if it works with RSVG.

**Reply by comsyspro** [reply]
i will try now to compile imagemagick with the option --with-rsvg=yes and will check if it will then work. i also saw that you can directly use command line of inkscape so this would be my second try. thank you.

**Reply by fmw42** [reply]
Thanks, snibgo. I did not know you could use -size with SVG. Is that only for Inkscape?

**Reply by snibgo** [reply]
It works for Inkscape. It doesn't work with the MSVG rasterizer. I don't know if it works with RSVG.

**Comment by comsyspro** [comment]
i just compiled IM  with the option --with-rsvg=yes.

now the delegates is shown which uses the syntax:
```
identify -list delegate | grep "svg ="
svg =>          "rsvg-convert' -o '%o' '%i"
```
but now the command causes the error
`/usr/local/bin/convert -size 2000x rectangle.svg rectangle.png`
convert: no images defined `rectangle.png' @ error/convert.c/ConvertImageCommand/3322.

when i use instead following command (-z 4 | zoom factor 4)
rsvg-convert -o rectangle.png -z 4 rectangle.svg
this is succesfully executed and the png is 4 times bigger

so why is there an error form the convert command?
`/usr/local/bin/convert -size 2000x rectangle.svg rectangle.png`

configuration after compiling
```
/usr/local/bin/convert -list configure

Path: /usr/local/lib/ImageMagick-7.0.11//config-Q16HDRI/configure.xml

Name                  Value
-------------------------------------------------------------------------------
CC                    gcc
CFLAGS                 -fopenmp -Wall -g -O2 -mtune=nehalem -fexceptions -pthread -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16
CODER_PATH            /usr/local/lib/ImageMagick-7.0.11/modules-Q16HDRI/coders
CONFIGURE             ./configure  '--with-bzlib=yes' '--with-jpeg=yes' '--with-png=yes' '--with-tiff=yes' '--with-xml=yes' '--with-rsvg=yes' '--with-modules'
CONFIGURE_PATH        /usr/local/etc/ImageMagick-7/
COPYRIGHT             Copyright (C) 1999-2021 ImageMagick Studio LLC
CPPFLAGS              -I/usr/local/include/ImageMagick-7
CXX                   g++
CXXFLAGS               -pthread
DEFS                  -DHAVE_CONFIG_H
DELEGATES             zlib
DISTCHECK_CONFIG_FLAGS  --disable-deprecated  --with-quantum-depth=16  --with-jemalloc=no  --with-umem=no  --with-autotrace=no  --with-fftw=no  --with-gslib=no  --with-fontpath=  --with-jxl=no  --with-perl=no
DOCUMENTATION_PATH    /usr/local/share/doc/ImageMagick-7
EXEC-PREFIX           /usr/local
EXECUTABLE_PATH       /usr/local/bin
FEATURES              DPC HDRI Cipher OpenMP Modules
FILTER_PATH           /usr/local/lib/ImageMagick-7.0.11/modules-Q16HDRI/filters
GIT_REVISION          18776
HOST                  x86_64-pc-linux-gnu
INCLUDE_PATH          /usr/local/include/ImageMagick-7
LDFLAGS               -L/usr/local/lib
LIB_VERSION           0x70B
LIB_VERSION_NUMBER    7,0,11,14
LIBRARY_PATH          /usr/local/lib/ImageMagick-7.0.11
LIBS                                -lz   -lltdl  -lm    -lpthread  -lgomp
NAME                  ImageMagick
PCFLAGS               -fopenmp -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16
PREFIX                /usr/local
QuantumDepth          16
RELEASE_DATE          2021-05-31
SHARE_PATH            /usr/local/share/ImageMagick-7
SHAREARCH_PATH        /usr/local/lib/ImageMagick-7.0.11/config-Q16HDRI
TARGET_CPU            x86_64
TARGET_OS             linux-gnu
TARGET_VENDOR         pc
VERSION               7.0.11
WEBSITE               https://imagemagick.org

Path: [built-in]

Name                  Value
-------------------------------------------------------------------------------
DELEGATES             ltdl zlib
FEATURES              Cipher DPC HDRI Modules OpenMP(4.5)
MAGICK_TEMPORARY_PATH /tmp
NAME                  ImageMagick
QuantumDepth          Q16
```

**Reply by fmw42** [reply]
You have to install the delegate before compiling IM with it.  Also I do not see any delegates other than zlib. So you need to install other delegates (PNG, TIFF, JPG etc) or install from binary.

**Reply by comsyspro** [reply]
i installed now some libraries png jpeg etc. and compiled again. 

now the command `/usr/local/bin/convert -size 2000x rectangle.svg rectangle.png` generally works and it converts the svg to the png but it doesn't size the png to the desired 2000px width. 

**so it seems that the size parameter doesn't work under linux.**
the delegation syntax of IM svg => "rsvg-convert' -o '%o' '%i" would need also parameters for -w width and -h height so the size parameter of IM could be used.

the new configuartion:
```
/usr/local/bin/convert -list configure

Path: /usr/local/lib/ImageMagick-7.0.11//config-Q16HDRI/configure.xml

Name                  Value
-------------------------------------------------------------------------------
CC                    gcc
CFLAGS                -I/usr/include/libxml2 -I/usr/include/libpng16    -Wall -g -O2 -mtune=nehalem -fexceptions -pthread -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16
CODER_PATH            /usr/local/lib/ImageMagick-7.0.11/modules-Q16HDRI/coders
CONFIGURE             ./configure  '--with-autotrace=no' '--with-gslib=yes' '--with-fontpath=' '--enable-delegate-build=yes' '--disable-shared' '--enable-static' '--with-modules=yes' '--without-perl' '--without-magick-plus-plus' '--disable-openmp' '--with-bzlib=yes' '--with-jpeg=yes' '--with-png=yes' '--with-tiff=yes' '--with-xml=yes' '--with-rsvg=yes'
CONFIGURE_PATH        /usr/local/etc/ImageMagick-7/
COPYRIGHT             Copyright (C) 1999-2021 ImageMagick Studio LLC
CPPFLAGS              -I/usr/local/include/ImageMagick-7
CXX                   g++
CXXFLAGS               -pthread
DEFS                  -DHAVE_CONFIG_H
DELEGATES             jbig jng jpeg lzma png ps tiff xml zlib
DISTCHECK_CONFIG_FLAGS  --disable-deprecated  --with-quantum-depth=16  --with-jemalloc=no  --with-umem=no  --with-autotrace=no  --with-fftw=no  --with-fontpath=  --with-jxl=no  --with-wmf=no  --with-perl=no
DOCUMENTATION_PATH    /usr/local/share/doc/ImageMagick-7
EXEC-PREFIX           /usr/local
EXECUTABLE_PATH       /usr/local/bin
FEATURES              DPC HDRI Cipher
FILTER_PATH           /usr/local/lib/ImageMagick-7.0.11/modules-Q16HDRI/filters
GIT_REVISION          18776
HOST                  x86_64-pc-linux-gnu
INCLUDE_PATH          /usr/local/include/ImageMagick-7
LDFLAGS               -L/usr/local/lib
LIB_VERSION           0x70B
LIB_VERSION_NUMBER    7,0,11,14
LIBRARY_PATH          /usr/local/lib/ImageMagick-7.0.11
LIBS                   -ljbig  -ltiff   -ljpeg    -lpng16 -lz                -llzma       -lxml2  -lz     -lm    -lpthread
NAME                  ImageMagick
PCFLAGS               -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16
PREFIX                /usr/local
QuantumDepth          16
RELEASE_DATE          2021-05-31
SHARE_PATH            /usr/local/share/ImageMagick-7
SHAREARCH_PATH        /usr/local/lib/ImageMagick-7.0.11/config-Q16HDRI
TARGET_CPU            x86_64
TARGET_OS             linux-gnu
TARGET_VENDOR         pc
VERSION               7.0.11
WEBSITE               https://imagemagick.org

Path: [built-in]

Name                  Value
-------------------------------------------------------------------------------
DELEGATES             jbig jng jpeg lzma png tiff xml zlib
FEATURES              Cipher DPC HDRI
MAGICK_TEMPORARY_PATH /tmp
NAME                  ImageMagick
QuantumDepth          Q16
```

**Reply by fmw42** [reply]
RSVG does not show in your list of delegates

DELEGATES             jbig jng jpeg lzma png ps tiff xml zlib

So I do not think it is installed with IM.   Check

`magick -list format`

and see what you have at the end of the line for SVG. Is it RSVG or MSVG/XML?

**Reply by comsyspro** [reply]
i will do it now with inkscape command line. IM is a little to complicated until it works as expected under linux. on windows it works fine.

i think the rsvg needs additional parameters for -w width and -h height because the convert command of IM is working now but it does not interpret the size parameter. perhaps you should add something like...
svg => "rsvg-convert' -o '%o' **-w '%w' -h '%h'** '%i"
... somehere in the source code of IM comparing to svg => "rsvg-convert' -o '%o' '%i"

**Comment by comsyspro** [comment]
finally i got it work. the solution was, that you need to install first all libraries for IM via `apt build-dep imagemagick`. after that using the option `./configure --with-rsvg=yes`. so i can confirm that IM can use the size parameter and works with RSVG. 

```
magick -list format | grep svg
RSVG* SVG       rw+   Librsvg SVG renderer (RSVG 2.48.9)
```

```
/usr/local/bin/convert -list configure

Path: /usr/local/lib/ImageMagick-7.0.11//config-Q16HDRI/configure.xml

Name                  Value
-------------------------------------------------------------------------------
CC                    gcc
CFLAGS                -I/usr/include/libxml2   -I/usr/include/cairo -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -pthread -I/usr/include/librsvg-2.0 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/cairo -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/libpng16  -I/usr/include/pango-1.0 -I/usr/include/fribidi -I/usr/include/harfbuzz -I/usr/include/cairo -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/OpenEXR  -I/usr/include/lqr-1 -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include  -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -pthread   -Wall -g -O2 -mtune=nehalem -fexceptions -pthread -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16
CODER_PATH            /usr/local/lib/ImageMagick-7.0.11/modules-Q16HDRI/coders
CONFIGURE             ./configure  '--with-autotrace=no' '--with-gslib=yes' '--with-fontpath=yes' '--enable-delegate-build=yes' '--enable-shared' '--enable-static' '--with-modules=yes' '--without-perl' '--without-magick-plus-plus' '--disable-openmp' '--with-bzlib=yes' '--with-jpeg=yes' '--with-png=yes' '--with-tiff=yes' '--with-xml=yes' '--with-rsvg=yes'
CONFIGURE_PATH        /usr/local/etc/ImageMagick-7/
COPYRIGHT             Copyright (C) 1999-2021 ImageMagick Studio LLC
CPPFLAGS              -I/usr/local/include/ImageMagick-7
CXX                   g++
CXXFLAGS               -pthread
DEFS                  -DHAVE_CONFIG_H
DELEGATES             bzlib djvu fontconfig freetype jbig jng jpeg lcms lqr lzma openexr pango png ps rsvg tiff webp wmf x xml zlib
DISTCHECK_CONFIG_FLAGS  --disable-deprecated  --with-quantum-depth=16  --with-jemalloc=no  --with-umem=no  --with-autotrace=no  --with-fftw=no  --with-fontpath=yes  --with-jxl=no  --with-perl=no
DOCUMENTATION_PATH    /usr/local/share/doc/ImageMagick-7
EXEC-PREFIX           /usr/local
EXECUTABLE_PATH       /usr/local/bin
FEATURES              DPC HDRI Cipher Modules
FILTER_PATH           /usr/local/lib/ImageMagick-7.0.11/modules-Q16HDRI/filters
GIT_REVISION          18776
HOST                  x86_64-pc-linux-gnu
INCLUDE_PATH          /usr/local/include/ImageMagick-7
LDFLAGS               -L/usr/local/lib
LIB_VERSION           0x70B
LIB_VERSION_NUMBER    7,0,11,14
LIBRARY_PATH          /usr/local/lib/ImageMagick-7.0.11
LIBS                   -llcms2 -lfreetype  -llqr-1 -lglib-2.0  -lxml2  -lfontconfig -lfreetype -lXext   -lSM -lICE -lX11  -lXt -lbz2 -lz   -lltdl  -lm    -lpthread
NAME                  ImageMagick
PCFLAGS               -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16
PREFIX                /usr/local
QuantumDepth          16
RELEASE_DATE          2021-05-31
SHARE_PATH            /usr/local/share/ImageMagick-7
SHAREARCH_PATH        /usr/local/lib/ImageMagick-7.0.11/config-Q16HDRI
TARGET_CPU            x86_64
TARGET_OS             linux-gnu
TARGET_VENDOR         pc
VERSION               7.0.11
WEBSITE               https://imagemagick.org

Path: [built-in]

Name                  Value
-------------------------------------------------------------------------------
DELEGATES             bzlib cairo djvu fontconfig freetype jbig jng jpeg lcms lqr ltdl lzma openexr pangocairo png rsvg tiff webp wmf x xml zlib
FEATURES              Cipher DPC HDRI Modules
MAGICK_TEMPORARY_PATH /tmp
NAME                  ImageMagick
QuantumDepth          Q16
```
